### PR TITLE
data: set up auto-generated API docs

### PR DESCRIPTION
**Pull request by oandregal**
**Related to** https://github.com/WordPress/gutenberg/issues/14227

This PR sets up auto-generated API docs for the `data` package through the `npm run docs:build` command and updates its README file.

## Test instructions

- Delete or modify the contents at `packages/data/README.md` within the start `<!-- START TOKEN(Autogenerated API docs) -->` and end `<!-- END TOKEN(Autogenerated API docs) -->` tokens.
- Execute `npm run docs:build` and verify they are re-created.
- Review that the API docs match the JSDoc of the exported symbols.

### CODE REVIEW

**Comment by gziolo** [comment]
I wouldn't consider it as a `Good First Review`. It should be reviewed by @aduth, @youknowriad or @coderkevin :)

File: packages/data/undocumented-exports.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Undocumented exports
```
Comment (gziolo) [inline packages/data/undocumented-exports.md]: > https://github.com/WordPress/gutenberg/pull/14271#discussion_r263735768

In that case, this needs to be delete as well.

Comment (nosolosw) [reply packages/data/undocumented-exports.md]: :man_facepalming: afd9249

File: packages/data/README.md [hunk 2]
```diff
@@ -227,45 +285,182 @@ registerStore( 'my-shop', {
...
+
+**Returns**
+
+`Object`: Registered store object.
+
+### RegistryConsumer
+
+[src/index.js#L14-L14](src/index.js#L14-L14)
+
+Undocumented declaration.
```
Comment (nosolosw) [inline packages/data/README.md]: This needs documenting. To be added as a JSDoc comment in the component.

File: packages/data/README.md [hunk 3]
```diff
@@ -227,45 +285,182 @@ registerStore( 'my-shop', {
...
+
+### RegistryProvider
+
+[src/index.js#L14-L14](src/index.js#L14-L14)
+
+Undocumented declaration.
```
Comment (nosolosw) [inline packages/data/README.md]: This needs documenting. To be added as a JSDoc comment in the component.

File: packages/data/README.md (hunk 2 above)
Comment (aduth) [reply packages/data/README.md]: Are you considering this a blocker? Do you need help with it?

Comment (nosolosw) [reply packages/data/README.md]: Not a blocker from my point of view. It was undocumented and will stay that way after all.

I feel this is ready and is a good addition, but would like feedback on 1) did I miss something? 2) does the general approach works well for this package or are things we should improve?

File: packages/data/README.md [hunk 4]
```diff
@@ -139,67 +139,125 @@ The `controls` option should be passed as an object where each key is the name o
...
+    -   In `@wordpress/data`, a `withSelect` mapping function can return `undefined` if it has no props to inject.
+-   In React Redux, the `mapDispatchToProps` argument can be defined as an object or a function.
+    -   In `@wordpress/data`, the `withDispatch` higher-order component creator must be passed a function.
 
-// Later, if necessary...
-unsubscribe();
-```
+## API
+
+It is very rare that you should access store methods directly. Instead, the following suite of functions and higher-order components is provided for the most common data access and manipulation needs.
```
Comment (aduth) [inline packages/data/README.md]: I'm struggling to remember the original context under which the phrasing was meant to apply: "It is very rare that you should access store methods directly" ([original document](https://github.com/WordPress/gutenberg/blob/f2bf8a1c66eb80cbf30c29c1cc00b08ad98bd9df/data/README.md#data-access-and-manipulation)). In any case, I think it reads even less appropriately now than ever, considering that the following set of API documentation are not really organized in any way, and are simply a listing of all exports from the module. I might suggest the whole paragraph be removed.

File: packages/data/src/index.js [hunk 5]
```diff
@@ -24,14 +24,104 @@ export { createRegistrySelector, createRegistryControl } from './factory';
...
+ * } );
+ * ```
+ *
  * @return {Function}       A reducer that invokes every reducer inside the reducers
  *                          object, and constructs a state object with the same shape.
  */
 export { combineReducers };
 
+/**
+ * Given the name of a registered store, returns an object of the store's selectors. The selector functions are been pre-bound to pass the current state automatically. As a consumer, you need only pass arguments of the selector, if applicable.
```
Comment (aduth) [inline packages/data/src/index.js]: Can we cap the line length here (and in other functions)? I'd suggest splitting it across 80 columns lines.

**Review by aduth — COMMENTED** [review]
I'm a bit more sensitive to this document than most other modules, since it's a key integration point for a majority of plugins. That said, I think while there's some general reorganization, it appears most of the examples are retained. Generally we might want to push toward more tutorials as part of the [handbook](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/javascript/), anyways.

File: packages/data/README.md (hunk 4 above)
Comment (nerrad) [reply packages/data/README.md]: As someone who has been a significant consumer of this api, I agree with @aduth's comment here.  Another thing I'm not sure about with regards to this documentation change is the order of things.  I think the part about generic stores should remain near the end of the documentation as it is a much more advanced api and thus would be referenced less often.

**Review by nerrad — COMMENTED** [review]
I'm a bit concerned about the arbitrary order of the documentation, but being that it's automated I understand it somewhat.  It'd be nice if the api can be grouped a bit more.  For example, general api functions that are used more frequently are `registerStore`, `dispatch`, `select`, `subscribe`, `withSelect` and `withDispatch`.  Then the others tend to be for more advanced usage.  With automation, could grouping be controlled a bit via special tags (i.e. `@docs:advanced`)?

File: packages/data/src/index.js (hunk 5 above)
Comment (nosolosw) [reply packages/data/src/index.js]: Done at 153a6b4

File: packages/data/README.md (hunk 4 above)
Comment (nosolosw) [reply packages/data/README.md]: Removed at 153a6b4

**Comment by nosolosw** [comment]
@aduth @nerrad Please, feel welcome to commit your suggestions to reorganization directly to this branch. I'm not familiar with this package as you are, so I'll need a little guidance.

Regarding grouping - for context, see [related discussion](https://github.com/WordPress/gutenberg/issues/14292#issuecomment-470224365). TLDR is: 

- we could switch to export order instead of alphabetical, but I don't think it'll provide better results. It takes a lot of energy to keep them well organized and up to date. As an example in point, these are the artifacts that aren't documented in this package: `createRegistry`, `createRegistryControl`, `createRegistrySelector`, `plugins`, `RegistryConsumer`, `RegistryProvider`, `use`, `withRegistry`.
- API docs work best when optimized for lookup, not onboarding (ie: alphabetical over most-common used first). We should create better onboarding docs in the handbook.

**Comment by nosolosw** [comment]
>  With automation, could grouping be controlled a bit via special tags (i.e. @docs:advanced)?

@nerrad I wanted to address this separately because that's a great question.

We currently have a very raw mechanism for grouping: we can update different parts of the README leveraging the `--ignore` and `--use-token` [options](https://github.com/WordPress/gutenberg/tree/master/packages/docgen#cli-options).

I don't think that's enough. I welcome new ideas and look forward to adding support for them! Some that I've already listed are: https://github.com/WordPress/gutenberg/issues/14333, https://github.com/WordPress/gutenberg/issues/14332 and https://github.com/WordPress/gutenberg/issues/14292 Note that these are generic and decoupled from a specific output. I'd want us to avoid adding metadata that is tied to a specific output (for example: adding metadata about which section in the README a symbol should be embedded, etc.).

**Comment by nosolosw** [comment]
@nerrad @aduth any chance you see this merged? I'd prefer to have these docs automated as automation is already showing how it helps to keep the docs up to date in the other packages. If you are strongly against this I'd rather close it so we can move on to other things.

**Review by nerrad — APPROVED** [review]
I say let's get this merged and then iterate.  Documentation is something that can be improved on over time.  I think there may be some benefit to getting this merged now being that there's some churn happening in wp.data that would be advantageous to get documented where possible.

**Pull request merged**